### ISSUE TITLE
Emulator closed with code: 10 and Emulator closed with code: 12

### ISSUE THREAD

**Issue by sgallese** [bzier/gym-mupen64plus#77]
This could be a mupen64plus issue and not relevant to the project.
Or possibly a configuration issue on my part.

Relevant error from stacktrace:
`Core Error: open_rom(): not a valid ROM image`
...
`Emulator closed with code: 10`

Cloned the master branch (90a4a9).
Added a Mario Kart n64 emulator named marioKart.n64 to the ROMs folder
Built the docker image according to README instructions.
Added the verifyEnv.py file to the root folder.

Appears emulator doesn't recognize the ROM.
However, I'm able to play the ROM without issue using OpenEmu.
Tried a different ROM, no luck.

Command
```
docker run -it \
  --name test-gym-env \
  -p 5900 \
 bz/gym-mupen64plus:0.0.5 `# This should match the image & tag you used during setup` \
  python -u gym-mupen64plus/verifyEnv.py
```
Note! I didn't use the mount code from the README, I was experiencing issues while using it. I assume adding the ROM to the proper folder before building the Docker image is sufficient though.

Output:
```
[2020-12-21 06:00:39,503] Making new env: Mario-Kart-Luigi-Raceway-v0
validate sub
('ControllerHTTPServer started on port ', 8082)
Initially on DISPLAY :0
Starting xvfb with command: ['Xvfb', ':0', '-screen', '0', '640x480x24', '-fbdir', '/dev/shm']

Using DISPLAY :0
Changed to DISPLAY :0
Starting emulator with comand: ['vglrun', '-d', ':0', 'mupen64plus', '--nospeedlimit', '--nosaveoptions', '--resolution', '640x480', '--gfx', 'mupen64plus-video-glide64.so', '--audio', 'dummy', '--input', '/usr/local/lib/mupen64plus/mupen64plus-input-bot.so', '/src/gym-mupen64plus/gym_mupen64plus/ROMs/marioKart.n64']
Calling mss.mss() with DISPLAY :0
 __  __                         __   _  _   ____  _
|  \/  |_   _ _ __   ___ _ __  / /_ | || | |  _ \| |_   _ ___
| |\/| | | | | '_ \ / _ \ '_ \| '_ \| || |_| |_) | | | | / __|
| |  | | |_| | |_) |  __/ | | | (_) |__   _|  __/| | |_| \__ \
|_|  |_|\__,_| .__/ \___|_| |_|\___/   |_| |_|   |_|\__,_|___/
             |_|         http://code.google.com/p/mupen64plus/
Mupen64Plus Console User-Interface Version 2.5.0

UI-Console: attached to core library 'Mupen64Plus Core' version 2.5.0
UI-Console:             Includes support for Dynamic Recompiler.
UI-Console:             Includes support for MIPS r4300 Debugger.
Core: Couldn't open configuration file '/root/.config/mupen64plus/mupen64plus.cfg'.  Using defaults.
Core Warning: No version number in 'Core' config section. Setting defaults.
Core Warning: No version number in 'CoreEvents' config section. Setting defaults.
UI-Console Warning: No version number in 'UI-Console' config section. Setting defaults.
Core Error: open_rom(): not a valid ROM image
UI-Console Error: core failed to open ROM image file '/src/gym-mupen64plus/gym_mupen64plus/ROMs/marioKart.n64'.
Changed back to DISPLAY :0
Emulator closed with code: 10
```

Also inspected the docker container, ensured that the ROM is placed in the correct location.
```
docker exec -t -i test-gym-env /bin/bash
root@e3d456090558:/src# cd gym-mupen64plus/gym_mupen64plus/ROMs/
root@e3d456090558:/src/gym-mupen64plus/gym_mupen64plus/ROMs# ls -l
total 8812
-rw-r--r-- 1 root root 9021826 Dec 21 05:12 marioKart.n64
```

I believe this error is bubbled up from the emulator:
https://github.com/mupen64plus/mupen64plus-core/blob/6cba5747cf1e870709c5d4aa1a954f2a0173255c/src/main/rom.c#L145

**Comment by bzier** [comment]
Hmm this is interesting. My first reaction was that the rom file probably didn't make it into the container, but then you answered that 🙂
My next thought is that the rom file itself is not compatible with mupen64plus. There are several [different formats](https://www.reddit.com/r/emulation/comments/7hrvzp/the_three_different_n64_rom_formats_explained_for/) for the rom files, and I'm wondering if there's one that openEmu can read that mupen64plus can't... I will send you the md5 sum for the rom I use. I'll also check what format it is.

**Comment by bzier** [comment]
The md5 sum of the ROM file I use is: `e19398a0fd1cc12df64fca7fbcaa82cc`
It was originally downloaded in a zip file with md5: `aa2b6cba6e78f15ae21d9b397b7f4273`

It appears that the one I'm using is the byteswapped 'v64' format ROM. From what I understand mupen64plus should support all of the formats, but who knows.

For what its worth, here are the first 64 bytes of the file:
![image](https://user-images.githubusercontent.com/714528/103020513-5fd2d780-44fd-11eb-94a3-c34754f94ced.png)

**Comment by bzier** [comment]
>Note! I didn't use the mount code from the README, I was experiencing issues while using it. I assume adding the ROM to the proper folder before building the Docker image is sufficient though.

That _should_ be sufficient. I'm curious though what issues you were having with the mount? What OS are you running on your host system?

**Comment by sgallese** [comment]
This is super helpful, @bzier, thank you.
I found a ROM with the right md5, that helped!
```
md5 ~/rom/marioKart.n64
MD5 (/Users/sgallese/rom/marioKart.n64) = e19398a0fd1cc12df64fca7fbcaa82cc
```

I was able to mount the directory as well!
Not sure what the problem was before, but works well.

I'm not facing a different error:
**Emulator closed with code: 12**
It can't find` /usr/local/lib/mupen64plus/mupen64plus-input-bot.so` but I verified the file exists 🤔

Investigating now. If any updates, I'll edit this comment.

**Building the docker image** 
```
➜  gym-mupen64plus git:(master) ✗ docker build -t bz/gym-mupen64plus:0.0.5 .
[+] Building 28.4s (19/19) FINISHED
 => [internal] load build definition from Dockerfile                                                         0.0s
 => => transferring dockerfile: 3.07kB                                                                       0.0s
 => [internal] load .dockerignore                                                                            0.0s
 => => transferring context: 78B                                                                             0.0s
 => [internal] load metadata for docker.io/library/ubuntu:xenial-20170915                                    2.0s
 => [auth] library/ubuntu:pull token for registry-1.docker.io                                                0.0s
 => [base 1/1] FROM docker.io/library/ubuntu:xenial-20170915@sha256:d45655633486615d164808b724b29406cb88e23  0.0s
 => [internal] load build context                                                                            0.0s
 => => transferring context: 13.39kB                                                                         0.0s
 => CACHED [stage-2 1/9] RUN apt-get update &&     apt-get install -y         python python-pip python-setu  0.0s
 => CACHED [stage-2 2/9] RUN pip install --upgrade pip                                                       0.0s
 => CACHED [stage-2 3/9] RUN wget "https://sourceforge.net/projects/virtualgl/files/2.5.2/virtualgl_2.5.2_a  0.0s
 => CACHED [buildstuff 1/3] RUN apt-get update &&     apt-get install -y         build-essential dpkg-dev l  0.0s
 => CACHED [buildstuff 2/3] WORKDIR /src/mupen64plus-src                                                     0.0s
 => CACHED [buildstuff 3/3] RUN git clone https://github.com/mupen64plus/mupen64plus-core &&         cd mup  0.0s
 => CACHED [stage-2 4/9] COPY --from=buildstuff /usr/local/lib/mupen64plus/mupen64plus-input-bot.so /usr/lo  0.0s
 => [stage-2 5/9] COPY . /src/gym-mupen64plus                                                                0.6s
 => [stage-2 6/9] COPY [./gym_mupen64plus/envs/Smash/smash.sra, /root/.local/share/mupen64plus/save/Super S  0.1s
 => [stage-2 7/9] WORKDIR /src/gym-mupen64plus                                                               0.0s
 => [stage-2 8/9] RUN pip install -e .                                                                      23.1s
 => [stage-2 9/9] WORKDIR /src                                                                               0.1s
 => exporting to image                                                                                       2.3s
 => => exporting layers                                                                                      2.3s
 => => writing image sha256:8b3a1eb7f4b7665fdf918a56b935449db57994cc706e4e9a18388c8ea7e2a842                 0.0s
 => => naming to docker.io/bz/gym-mupen64plus:0.0.5  
```

**Running the verifyEnv script**
```
➜  gym-mupen64plus git:(master) ✗ docker run -it \
   --name test-gym-env \
   -p 5900 \
   --mount source="/Users/sgallese/rom/",target=/src/gym-mupen64plus/gym_mupen64plus/ROMs,type=bind \
   bz/gym-mupen64plus:0.0.5 `# This should match the image & tag you used during setup` \
   python gym-mupen64plus/verifyEnv.py
[2020-12-24 01:47:09,939] Making new env: Mario-Kart-Luigi-Raceway-v0
validate sub
('ControllerHTTPServer started on port ', 8082)
Initially on DISPLAY :0
Starting xvfb with command: ['Xvfb', ':0', '-screen', '0', '640x480x24', '-fbdir', '/dev/shm']

Using DISPLAY :0
Changed to DISPLAY :0
Starting emulator with comand: ['vglrun', '-d', ':0', 'mupen64plus', '--nospeedlimit', '--nosaveoptions', '--resolution', '640x480', '--gfx', 'mupen64plus-video-glide64.so', '--audio', 'dummy', '--input', '/usr/local/lib/mupen64plus/mupen64plus-input-bot.so', '/src/gym-mupen64plus/gym_mupen64plus/ROMs/marioKart.n64']
Calling mss.mss() with DISPLAY :0
 __  __                         __   _  _   ____  _
|  \/  |_   _ _ __   ___ _ __  / /_ | || | |  _ \| |_   _ ___
| |\/| | | | | '_ \ / _ \ '_ \| '_ \| || |_| |_) | | | | / __|
| |  | | |_| | |_) |  __/ | | | (_) |__   _|  __/| | |_| \__ \
|_|  |_|\__,_| .__/ \___|_| |_|\___/   |_| |_|   |_|\__,_|___/
             |_|         http://code.google.com/p/mupen64plus/
Mupen64Plus Console User-Interface Version 2.5.0

UI-Console: attached to core library 'Mupen64Plus Core' version 2.5.0
UI-Console:             Includes support for Dynamic Recompiler.
UI-Console:             Includes support for MIPS r4300 Debugger.
Core: Couldn't open configuration file '/root/.config/mupen64plus/mupen64plus.cfg'.  Using defaults.
Core Warning: No version number in 'Core' config section. Setting defaults.
Core Warning: No version number in 'CoreEvents' config section. Setting defaults.
UI-Console Warning: No version number in 'UI-Console' config section. Setting defaults.
Core: Goodname: Mario Kart 64 (U) [!]
Core: Name: MARIOKART64
Core: MD5: 3A67D9986F54EB282924FCA4CD5F6DFF
Core: CRC: 3E5055B6 2E92DA52
Core: Imagetype: .v64 (byteswapped)
Core: Rom size: 12582912 bytes (or 12 Mb or 96 Megabits)
Core: Version: 1446
Core: Manufacturer: Nintendo
Core: Country: USA
UI-Console Status: Cheat codes disabled.
UI-Console: using Video plugin: 'Glide64 Video Plugin' v2.0.0
UI-Console: using Audio plugin: <dummy>
UI-Console Error: dlopen('/usr/local/lib/mupen64plus/mupen64plus-input-bot.so') failed: libjson-c.so.2: cannot open shared object file: No such file or directory
UI-Console Error: Specified Input plugin not found: /usr/local/lib/mupen64plus/mupen64plus-input-bot.so
Core Status: Rom closed.
Changed back to DISPLAY :0
Emulator closed with code: 12
```

**Verifying the .so file does exist**
```
docker exec -t -i test-gym-env /bin/bash
root@113d1d888e2b:/src# cd /usr/local/lib/mupen64plus/
root@113d1d888e2b:/usr/local/lib/mupen64plus# ls -l
total 12
-rw-r--r-- 1 root root 10336 Dec 20 06:23 mupen64plus-input-bot.so
```

**Comment by bzier** [comment]
It looks like it may actually be missing `libjson-c.so.2`, which is still a bit weird. That library should have been installed in one of the steps. Perhaps it isn't getting copied from the `buildstuff` layer (although I was sure I tested all this last year and had it working consistently). 

I'm hoping I'll get to dig in a bit more later this week/weekend. In the meantime, if you figure it out, please let me know.

**Comment by sgallese** [comment]
This is very helpful @bzier.
Once I created the docker image, I ran
`apt-get install libjson-c2 libjson-c-dev`
And now the verifyEnv.py script works.

**Comment by sgallese** [comment]
PR to ensure `libjson-c.so.2` makes it to the Docker image
https://github.com/bzier/gym-mupen64plus/pull/78

**Comment by bzier** [comment]
Closed by #78
